dev/com.ibm.ws.artifact.zip/src/com/ibm/ws/artifact/zip/internal/JavaArchive.java

```diff
@@ -0,0 +1,575 @@
+		
+		// The recorded size does not include the 12-byte header.   So let's add that
+		return specifiedZip64EoCDRecSize + ZIP64_EOCDR_HEADER_SIZE;
+	}
+		
+	/**
+	 * @param bytes   array of byte
+	 * @param offset  beginning of value in array of bytes
+	 * @param length  Must be <= 4
+	 * @return the bytes at the offset in reversed order
+	 */
+	private long getLittleEndianValue(byte[] bytes, int offset, int length) {
+		long value = 0;
+		for (int i = length - 1; i >= 0; i--) {
+			value = ((value << 8) | (bytes[offset + i] & 0xFF));
+		}
+		return value;
+	}
+	
+	/**
+	 * @return 4-byte signature using offset into byte array.
+	 */
+	private long getSignature(byte[] bytes, int offset) {
+		long value = 0;
+		for (int i = 0; i < 4; i++) {
+			value = ((value << 8) | (bytes[offset + i] & 0xFF));
+		}
+		return value;
+	}
+	
+	/**
+	 * @return true if the "total # of entries" offset in the End-of-Central-Directory-Record contains 0xFFFF
+	 */
+	private boolean isZip64() {
+		return (int) getLittleEndianValue(_endOfCentralDirectoryRecord, EOCDR_TOTAL_ENTRIES_OFFSET, 2) == ZIP64_INDICATOR;
+	}
+	
+	/**
+	 * Checks if the beginning of the actual archive, which might not be at the beginning of the
+	 * file contains the appropriate local-file-header-signature
+	 * @param file the zip file
+	 * @param archiveOffset the offset to the beginning of the archive within the file (Usually 0, but data might be attached in front of the archive).
+	 * @return true if valid
+	 * @throws IOException
+	 */
+	private boolean archiveStartsWithSignature(RandomAccessFile file, int archiveOffset) throws IOException {
+			
+		System.out.println("Archive offset is " + archiveOffset);
+		
+		file.seek(archiveOffset);
+		byte[] localFileHeaderSigBytes = new byte[ 4];
+		file.read(localFileHeaderSigBytes, 0, 4);
+		
+		if (isSignature(localFileHeaderSigBytes, 0, LOCAL_FILE_HEADER_SIGNATURE)) {
+			return true;
+		}
+		
+		return false;
+	}
+		
+	/**
+	 * @return 	 The location in the file where the archive actually starts. That will normally
+	 * be 0, but often scripts are pre-pended to the archive; for example to make
+	 * a self-extracting zip.
+	 * @throws IOException
+	 */
+	public long getActualArchiveOffset() throws IOException {
+		
+		// At offset 12 into the End-of-Central-Directory-Record is the size of the Central Directory 
+		long lengthOfCentralDirectory = getLittleEndianValue( _endOfCentralDirectoryRecord, EOCDR_CENTRAL_DIR_SIZE_OFFSET, 4 );
+		
+		// Offset 16 contains the relative offset from the beginning of the actual archive to the Central Directory
+		long offsetOfCentralDirectory = getLittleEndianValue( _endOfCentralDirectoryRecord, EOCDR_CENTRAL_DIR_OFFSET, 4 );
+		
+		long zip64RecLengths = isZip64() ? ( _zip64EoCDRecSize + ZIP64_LOCATOR_SIZE ) : 0;
+		
+		return _fileLength - _eocdRecordSize - zip64RecLengths - lengthOfCentralDirectory - offsetOfCentralDirectory;
+	}
+	
+	public static void main(String[] args) {
```

Comment:
This should probably be made into a unit test, with the test JARs constructed by the test, or stored remotely.

Reply:
Removed main for now, but yes there should be unit tests